src/KafkaNET.Library/Helper/KafkaSimpleManager.cs

```diff
@@ -177,31 +178,46 @@ public TopicMetadata RefreshMetadata(short versionId, string clientId, int corre
-                    int partitionCountInZK = GetTopicPartitionsFromZK(topic).Count;
-                    if (partitionCountInZK != partitionLeaders.Count)
-                        Logger.WarnFormat("RefreshMetadata exit return. Some partitions has no leader.  Topic:{0}  PartitionMetadata:{1} partitionLeaders:{2} != partitionCountInZK:{3}", topic, tempTopicMetadatas[topic].PartitionsMetadata.Count(), partitionLeaders.Count, partitionCountInZK);
+                    RefreshMetadataInternal(versionId, clientId, correlationId, topic, tempTopicMetadatas, tempTopicMetadatasLastUpdateTime, partitionLeaders);
+    
+                    if (tempTopicMetadatas.ContainsKey(topic))
+                    {
+                        this.TopicMetadatas[topic] = tempTopicMetadatas[topic];
+                        this.TopicMetadatasLastUpdateTime[topic] = tempTopicMetadatasLastUpdateTime[topic];
+                        this.TopicMetadataPartitionsLeaders[topic] = partitionLeaders;
+                        int partitionCountInZK = GetTopicPartitionsFromZK(topic).Count;
+                        if (partitionCountInZK != partitionLeaders.Count)
+                            Logger.WarnFormat("RefreshMetadata exit return. Some partitions has no leader.  Topic:{0}  PartitionMetadata:{1} partitionLeaders:{2} != partitionCountInZK:{3}", topic, tempTopicMetadatas[topic].PartitionsMetadata.Count(), partitionLeaders.Count, partitionCountInZK);
+                        else
+                            Logger.InfoFormat("RefreshMetadata exit return. Topic:{0}  PartitionMetadata:{1} partitionLeaders:{2} partitionCountInZK:{3}", topic, tempTopicMetadatas[topic].PartitionsMetadata.Count(), partitionLeaders.Count, partitionCountInZK);
+                        return this.TopicMetadatas[topic];
+                    }
                     else
-                        Logger.InfoFormat("RefreshMetadata exit return. Topic:{0}  PartitionMetadata:{1} partitionLeaders:{2} partitionCountInZK:{3}", topic, tempTopicMetadatas[topic].PartitionsMetadata.Count(), partitionLeaders.Count, partitionCountInZK);
-                    return this.TopicMetadatas[topic];
+                    {
+                        Logger.WarnFormat("Got null for metadata of topic {0}, will RecreateSyncProducerPoolForMetadata and retry . ", topic);
+                        RecreateSyncProducerPoolForMetadata();
+                    }
                 }
-                else
+                catch (Exception ex)
                 {
-                    Logger.WarnFormat("Got null for metadata of topic {0}, will RecreateSyncProducerPoolForMetadata and retry . ", topic);
+                    Logger.WarnFormat("Got exception while refreshing metadata of topic {0}, will RecreateSyncProducerPoolForMetadata and retry . {1} ",topic,
+                     ExceptionUtil.GetExceptionDetailInfo(ex));
                     RecreateSyncProducerPoolForMetadata();
```

Comment:
I dont think RecreateSyncProducerPoolForMetadata() should be extracted/moved outside the catch block as it should not need to be called each time RefreshMetadata() is called but only if RefreshMetadata() fail.

If what you mean is something such as setting a bool flag inside the catch block to indicate a failure and then call  RecreateSyncProducerPoolForMetadata() outside of the catch block if the flag was set, this could be done have would have the same intended result. but I am not sure I understand how this will make it better.

We could also add an extra try/catch around RecreateSyncProducerPoolForMetadata(); to honor the retryCount if you think this is better, but the retry loop was not really needed to start with since the caller of RefreshMetadata() could simply retry calling RefreshMetadata() if an exception is thrown.